pilot/pkg/bootstrap/server.go

```diff
@@ -899,35 +903,45 @@ func (s *Server) getCertKeyPaths(tlsOptions TLSOptions) (string, string) {
 	return key, cert
 }
 
-// getRootCertificates returns the root certificates from TLSOptions if available or from ca, and from SPIFFE bundle
-// endpoints.
-func (s *Server) getRootCertificates(tlsOptions TLSOptions) (*x509.CertPool, error) {
+// setPeerCertVerifier sets up a SPIFFE certificate verifier with the current istiod configuration.
+func (s *Server) setPeerCertVerifier(tlsOptions TLSOptions) error {
+	if tlsOptions.CaCertFile == "" && s.ca == nil && features.SpiffeBundleEndpoints == "" {
+		// Running locally without configured certs - no TLS mode
+		return nil
+	}
+	s.peerCertVerifier = spiffe.NewPeerCertVerifier()
 	var rootCertBytes []byte
 	var err error
 	if tlsOptions.CaCertFile != "" {
 		if rootCertBytes, err = ioutil.ReadFile(tlsOptions.CaCertFile); err != nil {
-			return nil, err
+			return err
 		}
-	} else {
+	} else if s.ca != nil {
 		rootCertBytes = s.ca.GetCAKeyCertBundle().GetRootCertPem()
 	}
-	cp := x509.NewCertPool()
-	cp.AppendCertsFromPEM(rootCertBytes)
 
-	if features.SpiffeBundlePaths != "" {
-		certMap, err := spiffe.RetrieveSpiffeBundleRootCertsFromStringInput(features.SpiffeBundlePaths, []*x509.Certificate{})
+	if len(rootCertBytes) != 0 {
+		block, _ := pem.Decode(rootCertBytes)
+		if block == nil {
+			return fmt.Errorf("failed to decode root cert PEM")
+		}
+		rootCert, err := x509.ParseCertificate(block.Bytes)
```

Comment:
Are we guaranteed that there is only one cert in rootCertBytes? Is it possible that there will be a cert pool of root certs?

Reply:
Good question. I think that's possible. I'll create a followup PR to fix this.